feat: update hardcode suffix,implements fallback logic,small refactor

luigibarbato opened the pull request:
Furyctl has a hardcoded suffix fury-Kubernetes- that adds to the generated URL for the download.

This suffix has a wrong convention, the right name of the Distribution is Kubernetes Fury Distribution. So the modules should also be Kubernetes Fury <Name> and their repositories kubernetes-fury-<name>, for consistency. The names in the readmes are already in this format.

This PR aims to update the hardcode suffix and implement a fallback logic that tries with the old suffix if the URL is not resolved to give a smooth transition to the users.

ralgozino reviewed (changes requested):
Hi @luigibarbato 

Sorry but the change in this PR is not working, this is what I did:

1. I downloaded a `Furyfile.yml` with the `init` command:

```bash
furyctl init --version v1.23.2
```

2. Added the `karrier` module to the bases, ending up with the following Furyfile.yml:

```yaml
# Copyright (c) 2022 SIGHUP s.r.l All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

versions:
  networking: v1.9.0
  monitoring: v1.14.2
  logging: v1.10.3
  ingress: v1.12.2
  dr: v1.9.2
  opa: v1.6.2
  karrier: main

bases:
  - name: networking
  - name: monitoring
  - name: logging
  - name: ingress
  - name: dr
  - name: opa
  - name: karrier
```

3. Then I run `furyctl vendor` to download the modules and I get this output:

```log
INFO[0000] using dev for package networking
INFO[0000] using dev for package monitoring
INFO[0000] using dev for package logging
INFO[0000] using dev for package ingress
INFO[0000] using dev for package dr
INFO[0000] using dev for package opa
INFO[0000] using dev for package karrier
INFO[0000] downloading: sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
INFO[0000] downloading: sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
INFO[0000] downloading: sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
INFO[0000] downloading: sighupio/kubernetes-fury-karrier/katalog?ref=main -> vendor/katalog/karrier
INFO[0000] downloading: sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
INFO[0000] downloading: sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
INFO[0000] downloading: sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
INFO[0002] cleaning git subfolder: vendor/katalog/karrier/.git
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-opa?ref=v1.6.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3649655918/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-ingress?ref=v1.12.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1478470050/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-monitoring?ref=v1.14.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3703353769/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-networking?ref=v1.9.0': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1484931747/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-logging?ref=v1.10.3': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1629662543/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-dr?ref=v1.9.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1521646596/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
```

If I now list the files inside the `vendor` folder, the only module downloaded is `karrier` (this module has the new -correct- name convention in the repo):

```bash
$ tree vendor -L 3
vendor
└── katalog
    └── karrier
        ├── control-plane
        ├── data-plane
        └── tests

5 directories, 0 files
```

I was expecting:
 1. no errors in the output but warnings when furyctl is doing the fallback
 2. all the listed modules are downloaded in the `vendor` folder.

ralgozino commented:
I forgot to add that we need a gopher paladin to review the code part of the PR, the current reviewers don't have enough Golang skills to do it :-)

luigibarbato commented:
> Hi @luigibarbato
> 
> Sorry but the change in this PR is not working, this is what I did:
> 
> 1. I downloaded a `Furyfile.yml` with the `init` command:
> 
> ```shell
> furyctl init --version v1.23.2
> ```
> 
> 2. Added the `karrier` module to the `bases` section, ending up with the following `Furyfile.yml`:
> 
> ```yaml
> # Copyright (c) 2022 SIGHUP s.r.l All rights reserved.
> # Use of this source code is governed by a BSD-style
> # license that can be found in the LICENSE file.
> 
> versions:
>   networking: v1.9.0
>   monitoring: v1.14.2
>   logging: v1.10.3
>   ingress: v1.12.2
>   dr: v1.9.2
>   opa: v1.6.2
>   karrier: main
> 
> bases:
>   - name: networking
>   - name: monitoring
>   - name: logging
>   - name: ingress
>   - name: dr
>   - name: opa
>   - name: karrier
> ```
> 
> 3. Then I run `furyctl vendor` to download the modules and I get this output:
> 
> ```
> INFO[0000] using dev for package networking
> INFO[0000] using dev for package monitoring
> INFO[0000] using dev for package logging
> INFO[0000] using dev for package ingress
> INFO[0000] using dev for package dr
> INFO[0000] using dev for package opa
> INFO[0000] using dev for package karrier
> INFO[0000] downloading: sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
> INFO[0000] downloading: sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
> INFO[0000] downloading: sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
> INFO[0000] downloading: sighupio/kubernetes-fury-karrier/katalog?ref=main -> vendor/katalog/karrier
> INFO[0000] downloading: sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
> INFO[0000] downloading: sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
> INFO[0000] downloading: sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
> INFO[0002] cleaning git subfolder: vendor/katalog/karrier/.git
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-opa?ref=v1.6.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3649655918/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-ingress?ref=v1.12.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1478470050/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-monitoring?ref=v1.14.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3703353769/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-networking?ref=v1.9.0': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1484931747/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-logging?ref=v1.10.3': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1629662543/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0002] error downloading 'ssh://git@github.com/sighupio/kubernetes-fury-dr?ref=v1.9.2': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter1521646596/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ```
> 
> If I now list the files inside the `vendor` folder, the only module downloaded is `karrier` (this module has the new -correct- name convention in the repo):
> 
> ```shell
> $ tree vendor -L 3
> vendor
> └── katalog
>     └── karrier
>         ├── control-plane
>         ├── data-plane
>         └── tests
> 
> 5 directories, 0 files
> ```
> 
> I was expecting:
> 
> 1. no errors in the output but warnings when furyctl is doing the fallback
> 2. all the listed modules are downloaded in the `vendor` folder.

I think I fixed it. the bug was only triggered in the case of ssh because trivially the error generated was different and not handled.

ralgozino commented:
> @ralgozino I think I fixed it. the bug was only triggered in the case of ssh because trivially the error generated was different and not handled.

Now seems to be working as expected 💪 

I've noticed 2 things though:

1. The first lines of the logs say:

```log
INFO[0000] using dev for package networking
INFO[0000] using dev for package monitoring
INFO[0000] using dev for package logging
INFO[0000] using dev for package ingress
INFO[0000] using dev for package dr
INFO[0000] using dev for package opa
[...]
```

when they should say:

```log
INFO[0000] using v1.9.0 for package networking
INFO[0000] using v1.14.2 for package monitoring
INFO[0000] using v1.10.3 for package logging
INFO[0000] using v1.12.2 for package ingress
INFO[0000] using v1.9.2 for package dr
INFO[0000] using v1.6.2 for package opa
[...]
```

this is a regression, you can test with furyctl 0.6.2 for example and see that it prints the versions instead of `dev`.

2. If you get an error, the exit code is still 0. You can test this for example using a random name for a base -something that doesn't exist- and then run `furyctl vendor`. Furyctl will try to download the non-existing module and fail with an error, but the exit code of the command is still 0. Can we fix this while we are at it? If it's not a simple fix we can work on it later.

ralgozino reviewed (changes requested):
logs should say version being downloaded (as before) instead of `dev`

luigibarbato commented:
> > @ralgozino I think I fixed it. the bug was only triggered in the case of ssh because trivially the error generated was different and not handled.
> 
> Now seems to be working as expected 💪
> 
> I've noticed 2 things though:
> 
> 1. The first lines of the logs say:
> 
> ```
> INFO[0000] using dev for package networking
> INFO[0000] using dev for package monitoring
> INFO[0000] using dev for package logging
> INFO[0000] using dev for package ingress
> INFO[0000] using dev for package dr
> INFO[0000] using dev for package opa
> [...]
> ```
> 
> when they should say:
> 
> ```
> INFO[0000] using v1.9.0 for package networking
> INFO[0000] using v1.14.2 for package monitoring
> INFO[0000] using v1.10.3 for package logging
> INFO[0000] using v1.12.2 for package ingress
> INFO[0000] using v1.9.2 for package dr
> INFO[0000] using v1.6.2 for package opa
> [...]
> ```
> 
> this is a regression, you can test with furyctl 0.6.2 for example and see that it prints the versions instead of `dev`.
> 
> 2. If you get an error, the exit code is still 0. You can test this for example using a random name for a base -something that doesn't exist- and then run `furyctl vendor`. Furyctl will try to download the non-existing module and fail with an error, but the exit code of the command is still 0. Can we fix this while we are at it? If it's not a simple fix we can work on it later.

@ralgozino It should be ok now :)

ralgozino reviewed (changes requested):
Hello @luigibarbato !

Now the version is correctly printed but the exit code is still zero in the case of an error.

I've found another case that we need to handle:

If the download of a module with the new format in the name success but fails with the old one, you still get an error for the old format. The fallback URL should not be even tried in this case.

You can test this with the module `policy`: it gets downloaded with the new format but errors with the old one.

Here's some example output:

```bash
$ furyctl vendor
INFO[0000] using v1.9.0 for package networking
INFO[0000] using v1.14.2 for package monitoring
INFO[0000] using v1.10.3 for package logging
INFO[0000] using v1.12.2 for package ingress
INFO[0000] using v1.9.2 for package dr
INFO[0000] using v1.6.2 for package opa
INFO[0000] downloading: sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
INFO[0000] downloading: sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
INFO[0000] downloading: sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
INFO[0000] downloading: sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
INFO[0000] downloading: sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
INFO[0000] downloading: sighupio/fury-kubernetes-non-existent/katalog?ref= -> vendor/katalog/non-existent
INFO[0000] downloading: sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
INFO[0000] downloading: sighupio/fury-kubernetes-policy/katalog?ref= -> vendor/katalog/policy
INFO[0001] removing target path: vendor/katalog/opa
INFO[0001] cleaning git subfolder: vendor/katalog/opa/.git
INFO[0001] removing target path: vendor/katalog/dr
INFO[0001] cleaning git subfolder: vendor/katalog/dr/.git
INFO[0002] removing target path: vendor/katalog/logging
INFO[0002] cleaning git subfolder: vendor/katalog/logging/.git
INFO[0002] removing target path: vendor/katalog/networking
INFO[0002] cleaning git subfolder: vendor/katalog/networking/.git
INFO[0002] removing target path: vendor/katalog/monitoring
INFO[0002] cleaning git subfolder: vendor/katalog/monitoring/.git
INFO[0003] removing target path: vendor/katalog/ingress
INFO[0003] cleaning git subfolder: vendor/katalog/ingress/.git
ERRO[0003] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-policy?ref=': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter798862759/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0003] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3018795903/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.

$ echo $?
0
```

luigibarbato commented:
> Hello @luigibarbato !
> 
> Now the version is correctly printed but the exit code is still zero in the case of an error.
> 
> I've found another case that we need to handle:
> 
> If the download of a module with the new format in the name success but fails with the old one, you still get an error for the old format. The fallback URL should not be even tried in this case.
> 
> You can test this with the module `policy`: it gets downloaded with the new format but errors with the old one.
> 
> Here's some example output:
> 
> ```shell
> $ furyctl vendor
> INFO[0000] using v1.9.0 for package networking
> INFO[0000] using v1.14.2 for package monitoring
> INFO[0000] using v1.10.3 for package logging
> INFO[0000] using v1.12.2 for package ingress
> INFO[0000] using v1.9.2 for package dr
> INFO[0000] using v1.6.2 for package opa
> INFO[0000] downloading: sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
> INFO[0000] downloading: sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
> INFO[0000] downloading: sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
> INFO[0000] downloading: sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
> INFO[0000] downloading: sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
> INFO[0000] downloading: sighupio/fury-kubernetes-non-existent/katalog?ref= -> vendor/katalog/non-existent
> INFO[0000] downloading: sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
> INFO[0000] downloading: sighupio/fury-kubernetes-policy/katalog?ref= -> vendor/katalog/policy
> INFO[0001] removing target path: vendor/katalog/opa
> INFO[0001] cleaning git subfolder: vendor/katalog/opa/.git
> INFO[0001] removing target path: vendor/katalog/dr
> INFO[0001] cleaning git subfolder: vendor/katalog/dr/.git
> INFO[0002] removing target path: vendor/katalog/logging
> INFO[0002] cleaning git subfolder: vendor/katalog/logging/.git
> INFO[0002] removing target path: vendor/katalog/networking
> INFO[0002] cleaning git subfolder: vendor/katalog/networking/.git
> INFO[0002] removing target path: vendor/katalog/monitoring
> INFO[0002] cleaning git subfolder: vendor/katalog/monitoring/.git
> INFO[0003] removing target path: vendor/katalog/ingress
> INFO[0003] cleaning git subfolder: vendor/katalog/ingress/.git
> ERRO[0003] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-policy?ref=': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter798862759/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0003] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter3018795903/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> 
> $ echo $?
> 0
> ```

Hi @ralgozino, thanks to @lzecca78 we found several problems in the implementation of the goroutines causing unexpected errors. Now not only should you no longer experience any unexpected problems, but the code should be slightly more stable. :)

Al-Pragliola commented on cmd/download.go:
```diff
@@ -7,150 +7,252 @@ package cmd
+	}
+}
+
+//getConsumableURL returns an url that can be used for download
```
nitpick
```suggestion
// getConsumableURL returns an url that can be used for download
```

Al-Pragliola commented on cmd/download.go:
```diff
@@ -7,150 +7,252 @@ package cmd
+
+func (n *PackageURL) getURLFromCompanyRepos() string {
+	if len(n.Blocks) == 0 {
+		// todo should return error?
```
maybe we should reference with TODO in a task

luigibarbato replied:
Maybe, we'll discuss it.

Al-Pragliola commented on cmd/download.go:
```diff
@@ -7,150 +7,252 @@ package cmd
+	if len(n.Blocks) == 1 {
+		return fmt.Sprintf("%s-%s%s//%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, n.Version)
+	}
+	// always len(n.Blocks) >= 2 {
```
still needed?

luigibarbato replied:
I think not but it's a legacy code so I would leave it for now.

ralgozino reviewed (changes requested):
Hello!

Some feedback after testing the latest commits:

1. furyctl is now showing an error if the version is not specified. This is a breaking change in the behaviour, previously it would default to the main branch, right? 

```bash
$ furyctl vendor
ERRO[0000] no version found for package policy. Read the docs in the errors section
$ echo $?
0
```

Where are these docs that the error sends me to read? The error is a little harsh on the user.

I'd rather not make that breaking change.

Notice that even though I got an error the exit code is still zero.

2. If I put the version in the Furyfile I get around the no version error and got this output for my test case:

```bash
$ furyctl vendor
INFO[0000] using v1.9.0 for package networking
INFO[0000] using v1.14.2 for package monitoring
INFO[0000] using v1.10.3 for package logging
INFO[0000] using v1.12.2 for package ingress
INFO[0000] using v1.9.2 for package dr
INFO[0000] using v1.6.2 for package opa
INFO[0000] using v1.6.2 for package policy
INFO[0000] using master for package non-existent
INFO[0000] downloading: sighupio/kubernetes-fury-non-existent/katalog?ref=master -> vendor/katalog/non-existent
INFO[0000] downloading: sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
INFO[0000] downloading: sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
INFO[0000] downloading: sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
INFO[0000] downloading: sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
INFO[0000] downloading: sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
INFO[0000] downloading: sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
INFO[0000] downloading: sighupio/kubernetes-fury-policy/katalog?ref=v1.6.2 -> vendor/katalog/policy
WARN[0001] error downloading sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2, falling back to sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2
INFO[0001] downloading: sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
WARN[0001] error downloading sighupio/kubernetes-fury-non-existent/katalog?ref=master, falling back to sighupio/fury-kubernetes-non-existent/katalog?ref=master
INFO[0001] downloading: sighupio/fury-kubernetes-non-existent/katalog?ref=master -> vendor/katalog/non-existent
WARN[0001] error downloading sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2, falling back to sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2
INFO[0001] downloading: sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
WARN[0001] error downloading sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2, falling back to sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2
INFO[0001] downloading: sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
WARN[0001] error downloading sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0, falling back to sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0
INFO[0001] downloading: sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
WARN[0001] error downloading sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2, falling back to sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2
INFO[0001] downloading: sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
WARN[0001] error downloading sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3, falling back to sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3
INFO[0001] downloading: sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
INFO[0001] cleaning git subfolder: vendor/katalog/policy/.git
INFO[0002] cleaning git subfolder: vendor/katalog/opa/.git
INFO[0003] cleaning git subfolder: vendor/katalog/networking/.git
INFO[0003] cleaning git subfolder: vendor/katalog/ingress/.git
INFO[0004] cleaning git subfolder: vendor/katalog/dr/.git
INFO[0004] cleaning git subfolder: vendor/katalog/monitoring/.git
INFO[0005] cleaning git subfolder: vendor/katalog/logging/.git
ERRO[0005] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
ERRO[0005] ERROR DOWNLOADING                             error="error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
Error: error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'...
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Is it my impression or the logs are a little mangled? 
The error message get's printed 3 times, making the output difficult to read. Can we improve the formatting?
I can understand that because of reasons we print >1 time the error, but the final one should be clear to the user.

luigibarbato commented on cmd/download.go:
```diff
@@ -7,150 +7,252 @@ package cmd
+	return humanReadableSrc
 }
 
 func removeDir(dir string) error {
```
I think the same. But it's a legacy code so I would leave it for now.

luigibarbato commented:
> Hello!
> 
> Some feedback after testing the latest commits:
> 
> 1. furyctl is now showing an error if the version is not specified. This is a breaking change in the behaviour, previously it would default to the main branch, right?
> 
> ```shell
> $ furyctl vendor
> ERRO[0000] no version found for package policy. Read the docs in the errors section
> $ echo $?
> 0
> ```
> 
> Where are these docs that the error sends me to read? The error is a little harsh on the user.
> 
> I'd rather not make that breaking change.
> 
> Notice that even though I got an error the exit code is still zero.
> 
> 2. If I put the version in the Furyfile I get around the no version error and got this output for my test case:
> 
> ```shell
> $ furyctl vendor
> INFO[0000] using v1.9.0 for package networking
> INFO[0000] using v1.14.2 for package monitoring
> INFO[0000] using v1.10.3 for package logging
> INFO[0000] using v1.12.2 for package ingress
> INFO[0000] using v1.9.2 for package dr
> INFO[0000] using v1.6.2 for package opa
> INFO[0000] using v1.6.2 for package policy
> INFO[0000] using master for package non-existent
> INFO[0000] downloading: sighupio/kubernetes-fury-non-existent/katalog?ref=master -> vendor/katalog/non-existent
> INFO[0000] downloading: sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
> INFO[0000] downloading: sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
> INFO[0000] downloading: sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
> INFO[0000] downloading: sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
> INFO[0000] downloading: sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
> INFO[0000] downloading: sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
> INFO[0000] downloading: sighupio/kubernetes-fury-policy/katalog?ref=v1.6.2 -> vendor/katalog/policy
> WARN[0001] error downloading sighupio/kubernetes-fury-monitoring/katalog?ref=v1.14.2, falling back to sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2
> INFO[0001] downloading: sighupio/fury-kubernetes-monitoring/katalog?ref=v1.14.2 -> vendor/katalog/monitoring
> WARN[0001] error downloading sighupio/kubernetes-fury-non-existent/katalog?ref=master, falling back to sighupio/fury-kubernetes-non-existent/katalog?ref=master
> INFO[0001] downloading: sighupio/fury-kubernetes-non-existent/katalog?ref=master -> vendor/katalog/non-existent
> WARN[0001] error downloading sighupio/kubernetes-fury-dr/katalog?ref=v1.9.2, falling back to sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2
> INFO[0001] downloading: sighupio/fury-kubernetes-dr/katalog?ref=v1.9.2 -> vendor/katalog/dr
> WARN[0001] error downloading sighupio/kubernetes-fury-ingress/katalog?ref=v1.12.2, falling back to sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2
> INFO[0001] downloading: sighupio/fury-kubernetes-ingress/katalog?ref=v1.12.2 -> vendor/katalog/ingress
> WARN[0001] error downloading sighupio/kubernetes-fury-networking/katalog?ref=v1.9.0, falling back to sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0
> INFO[0001] downloading: sighupio/fury-kubernetes-networking/katalog?ref=v1.9.0 -> vendor/katalog/networking
> WARN[0001] error downloading sighupio/kubernetes-fury-opa/katalog?ref=v1.6.2, falling back to sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2
> INFO[0001] downloading: sighupio/fury-kubernetes-opa/katalog?ref=v1.6.2 -> vendor/katalog/opa
> WARN[0001] error downloading sighupio/kubernetes-fury-logging/katalog?ref=v1.10.3, falling back to sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3
> INFO[0001] downloading: sighupio/fury-kubernetes-logging/katalog?ref=v1.10.3 -> vendor/katalog/logging
> INFO[0001] cleaning git subfolder: vendor/katalog/policy/.git
> INFO[0002] cleaning git subfolder: vendor/katalog/opa/.git
> INFO[0003] cleaning git subfolder: vendor/katalog/networking/.git
> INFO[0003] cleaning git subfolder: vendor/katalog/ingress/.git
> INFO[0004] cleaning git subfolder: vendor/katalog/dr/.git
> INFO[0004] cleaning git subfolder: vendor/katalog/monitoring/.git
> INFO[0005] cleaning git subfolder: vendor/katalog/logging/.git
> ERRO[0005] error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'... ERROR: Repository not found. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.
> ERRO[0005] ERROR DOWNLOADING                             error="error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
> Error: error downloading 'ssh://git@github.com/sighupio/fury-kubernetes-non-existent?ref=master': /usr/local/bin/git exited with 128: Cloning into '/var/folders/2d/79kvs5v151l4wnh6mllfrv5m0000gn/T/getter872312697/temp'...
> ERROR: Repository not found.
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.
> ```
> 
> Is it my impression or the logs are a little mangled? The error message get's printed 3 times, making the output difficult to read. Can we improve the formatting? I can understand that because of reasons we print >1 time the error, but the final one should be clear to the user.

Okay. Rami, first of all, thank you for the excellent suggestions and forgive me for the lack of attention. 

1. Following a note from @lzecca78 , we realized that there is a limitation of the tool: when the user declares the name of a module within the furyfile and omits the version, furyctl still attempts to download it. I totally agree with you, the implementation was rushed and without taking into consideration the many many variables involved. it has been removed 

2. Unconsciously, I thought it was an expected result. In fact, until now furyctl was not returning any errors and so the only output it returned was the one handled internally by us. Cobra, the library we use for the implementation part of the command line, turns on by default the option to print any error generated by the specific command. The option has been disabled and the format of the logs should match the expected one.

luigibarbato commented on cmd/vendor.go:
```diff
@@ -11,18 +11,27 @@ import (
+	Use:           "vendor",
+	Short:         "Download dependencies specified in Furyfile.yml",
+	Long:          "Download dependencies specified in Furyfile.yml",
+	SilenceUsage:  true,
+	SilenceErrors: true,
```
It's needed because we could have duplicated errors caused by internal logs

omissis replied:
I'd rather have duplicated logs and fix them rather than throwing stuff under the carpet, but I need to try the command to get a better grasp on this

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
 	}
-	return nil
+
+	return errors.New("download failed. See the logs for more information")
```
This line will have this function always return error, why is that?

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
+
+	// Closing the job channel
 	close(jobs)
 	logrus.Debugf("closed jobs")
```
please remove, this is not a useful debug information for the end user.

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
+
+	// Closing the error channel
 	close(errChan)
+	logrus.Debugf("closed errChan")
```
please remove, this is not a useful debug information for the end user.

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
 	close(errChan)
+	logrus.Debugf("closed errChan")
+
 	logrus.Debugf("finished")
```
```suggestion
	logrus.Debugf("finished downloading all packages")
```

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
+	logrus.Debugf("workers = %d", len(jobs))
+
+	// Starting the workers to download the packages in parallel
+	for i, data := range packages {
```
I find the previous approach (that used numberOfWorkers) to be preferable as:
- it allowed the user to choose between parallel and serial execution
- it took available CPUs into account instead of fanning out without control

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
+
+	// Checking git clone protocol
+	p := ""
+	if opts.Https {
+		p = httpsRepoPrefix
 	} else {
-		numberOfWorkers = 1
+		p = sshRepoPrefix
+	}
```
p can be instantiead as `sshRepoPrefix`, so that `else` is not needed anymore

omissis commented on cmd/download.go:
```diff
@@ -5,166 +5,258 @@
+
+		downloadErr = get(pU.getConsumableURL(), data.Dir, getter.ClientModeDir, true)
+		if downloadErr != nil {
+			err := os.RemoveAll(data.Dir)
+			if err != nil {
+				logrus.Errorf("error removing directory %s: %s", data.Dir, err.Error())
+			}
```
```suggestion
			if err := os.RemoveAll(data.Dir); err != nil {
				logrus.Errorf("error removing directory %s: %s", data.Dir, err.Error())
			}
```

omissis replied:
why is this error logged here and the downloadErr passed through an error channel and logger later?

omissis reviewed (changes requested):
I tried the command as highlighted in @ralgozino's comment https://github.com/sighupio/furyctl/pull/86#pullrequestreview-1052479037 and the process hanged as shown in the screenshot below (I assume it's an issue with go routines and/or wait groups).
<img width="1153" alt="Screenshot 2022-08-02 at 09 04 42" src="https://user-images.githubusercontent.com/197604/182313766-4bbd6229-7b48-4c8a-8e69-7a38d1aaa7ee.png">

Besides that issue that will likely require more investigation, the rest of the code is ok for me.

luigibarbato commented:
> I tried the command as highlighted in @ralgozino's comment [#86 (review)](https://github.com/sighupio/furyctl/pull/86#pullrequestreview-1052479037) and the process hanged as shown in the screenshot below (I assume it's an issue with go routines and/or wait groups). <img alt="Screenshot 2022-08-02 at 09 04 42" width="1153" src="https://user-images.githubusercontent.com/197604/182313766-4bbd6229-7b48-4c8a-8e69-7a38d1aaa7ee.png">
> 
> Besides that issue that will likely require more investigation, the rest of the code is ok for me.

Fixed, test again please

ralgozino reviewed (changes requested):
This still has issues:
- If a version is not specified you get an empty space in the `INFO` log
- If you try to download a private repository with `-H` (unauthenticated) it fails silently - the module is not downloaded but furyctl doesn't fail.
- If you put a wrong version, the same as the previous

I've created some "test cases" so you can test yourself
[furyctl-vendor-fallback-tests.zip](https://github.com/sighupio/furyctl/files/9249994/furyctl-vendor-fallback-tests.zip)

luigibarbato commented:
> This still has issues:
> 
> * If a version is not specified you get an empty space in the `INFO` log
> * If you try to download a private repository with `-H` (unauthenticated) it fails silently - the module is not downloaded but furyctl doesn't fail.
> * If you put a wrong version, the same as the previous
> 
> I've created some "test cases" so you can test yourself [furyctl-vendor-fallback-tests.zip](https://github.com/sighupio/furyctl/files/9249994/furyctl-vendor-fallback-tests.zip)

as discussed, are unleashed problems inherent to the above PR. Let's track and fix them in separate tasks :)

ralgozino reviewed (changes requested):
> > This still has issues:
> > 
> > * If a version is not specified you get an empty space in the `INFO` log
> > * If you try to download a private repository with `-H` (unauthenticated) it fails silently - the module is not downloaded but furyctl doesn't fail.
> > * If you put a wrong version, the same as the previous
> > 
> > I've created some "test cases" so you can test yourself [furyctl-vendor-fallback-tests.zip](https://github.com/sighupio/furyctl/files/9249994/furyctl-vendor-fallback-tests.zip)
> 
> as discussed, are unleashed problems inherent to the above PR. Let's track and fix them in separate tasks :)

Hi @luigibarbato 

I've run again the tests with furyctl `0.7.0`, `master` and with this `PR`. I got the same results with `master` than with `0.7.0`, but the issues I've mentioned in my previous comment are introduced in this `PR`.

Next, the results of the tests I've done:

**Version not specified**
When you don't specify the version of a `base` in the Furyfile...

- In `master`: it just doesn't print any information when there's no version (I consider this a bug in `master`)
- In `PR`: it prints the same message of the bases with a version but with an empty space instead of the version.

I've pushed a fix.

**Private Repo**
when you use as a `base` a private repository in the Furyfile and run `furyctl vendor -H` (uses HTTP so without authenticating against GitHub)...

- In `master`: you get an error printed and exit code zero (the exit code is "wrong")
- In `PR`: you don't get any error in the logs and exit code zero.

if you run `vendor` instead, you get the expected output:

- In `master`: it downloads the repo.
- In `PR`: it downloads the repo.

This PR is introducing a bug of failing silently when you try to download a repository that needs authentication and the `-H` flag at the same time.

**Wrong version**
When you put a version that does not exists for a `base` in the Furyfile and run `vendor` (with and without `-H`)...

- In `master`: you get an error in the logs in both cases and exit code zero (the exit code is "wrong")
- In `PR`: you don't get any error and exit code zero.

This PR is introducing a bug of failing silently when I try to download a version that does not exist.

---

So, these two issues are inherent to this PR and we need to solve them before merging.

On the good side, the repo name fallback is working properly :-)

Thank you for your patience

ralgozino commented:
BTW: Linting is not passing anymore, I suspect that it started breaking now because the golang container was not tagged in the lining step of the `drone.yml`, I believe @omissis has the fix ready locally.

ralgozino reviewed (approved):
tested with latest fix by @Al-Pragliola , lgmt 🎉
